proof/refine/Invariants_H.thy

```diff
@@ -1800,8 +1485,70 @@ lemma obj_at'_pspaceI:
   "obj_at' t ref s \<Longrightarrow> ksPSpace s = ksPSpace s' \<Longrightarrow>  obj_at' t ref s'"
   by (auto intro!: projectKO_stateI simp: obj_at'_def ps_clear_def)
 
+lemma valid_untyped'_pspaceI:
+  "\<lbrakk>ksPSpace s = ksPSpace s'; valid_untyped' d p n idx s\<rbrakk>
+  \<Longrightarrow> valid_untyped' d p n idx s'"
+  by (simp add: valid_untyped'_def ko_wp_at'_def ps_clear_def)
+
+lemma typ_at'_pspaceI:
+  "typ_at' T p s \<Longrightarrow> ksPSpace s = ksPSpace s' \<Longrightarrow> typ_at' T p s'"
+  by (simp add: typ_at'_def ko_wp_at'_def ps_clear_def)
+
+(* all architectures have same slots, only cteSizeBits will vary *)
+lemma (in Arch) tcb_cte_cases_distinct_n:
+  "distinct [
+     0 << cteSizeBits :: machine_word,
+     1 << cteSizeBits,
+     2 << cteSizeBits,
+     3 << cteSizeBits,
+     4 << cteSizeBits]"
+  by (simp add: cteSizeBits_def wordSizeCase_def wordBits_def word_size)
+
+requalify_facts Arch.tcb_cte_cases_distinct_n
+
+lemmas tcb_cte_cases_neqs_n =
+  tcb_cte_cases_distinct_n[simplified]
+  distinct_rev[THEN iffD2, OF tcb_cte_cases_distinct_n, simplified]
+
+lemma tcb_cte_cases_simps[simp]:
+  "tcb_cte_cases 0  = Some (tcbCTable, tcbCTable_update)"
+  "tcb_cte_cases (1 << cteSizeBits) = Some (tcbVTable, tcbVTable_update)"
+  "tcb_cte_cases (2 << cteSizeBits) = Some (tcbReply, tcbReply_update)"
+  "tcb_cte_cases (3 << cteSizeBits) = Some (tcbCaller, tcbCaller_update)"
+  "tcb_cte_cases (4 << cteSizeBits) = Some (tcbIPCBufferFrame, tcbIPCBufferFrame_update)"
+  (* FIXME arch-split: warning due to tcb_cte_cases_neqs overlapping *)
+  by (simp add: tcb_cte_cases_neqs_n tcb_cte_cases_def)+
+
+lemmas tcbSlot_defs = tcbCTableSlot_def tcbVTableSlot_def tcbReplySlot_def tcbCallerSlot_def
+                      tcbIPCBufferSlot_def
+
+lemma tcb_cte_cases_distinct:
+  "distinct [
+     tcbCTableSlot << cteSizeBits :: machine_word,
+     tcbVTableSlot << cteSizeBits,
+     tcbReplySlot << cteSizeBits,
+     tcbCallerSlot << cteSizeBits,
+     tcbIPCBufferSlot << cteSizeBits]"
+  (* FIXME arch-split: warning due to tcb_cte_cases_neqs overlapping *)
+  by (simp add: tcbSlot_defs tcb_cte_cases_neqs_n)
```

Comment:
Cool. In the distant future, this could potentially be the form that we do not unfold further in other proofs and that the tcb cases then use everywhere. With that, there would be at least some hope that it is symbolic enough to not reduce to the same terms in `tcb_cte_cases_neqs` and therefore get rid of the warning (and the additional simplification work Isabelle has to do in each proof).

Reply:
I'm not following by what you mean here. If you mean that we don't unfold the `tcbCTableSlot` etc so we end up only needing these versions, then that would be nice, and would indeed not cause duplication in the `_neqs` lemma, which does normalise things around `0` and `1`. Don't think `Distinct_Cmd` will help much with that, it will either give a result not in simp normal form, or if we use `[simplified]`, duplicates as before.

Reply:
I pushed a fix for this, which also removes the FIXMEs. Playing around with `Distinct_Cmd` didn't really improve over `distinct []` and your attribute hack, but I guess it in theory could work around the attribute hack if it wasn't so awkward to use.

Reply:
> If you mean that we don't unfold the tcbCTableSlot etc

Yup, that's what I meant. I don't know where exactly the unfoldings come in, but in old code it could already be at the spec level. In any case, that's for another time.

Nice work on removing the duplicates!